base/src/main/scala/co/uproot/abandon/Ast.scala

```diff
@@ -59,6 +59,135 @@ case class Date(year: Int, month: Int, day: Int) {
   }
 }
 
+/**
+ * Trait for stackable Transaction filters
+ * These filters operate on "raw" AST level,
+ * so all filtering decisions are based on unprocessed
+ * information.
+ *
+ * TxnFilterStack is used to glue these filters together.
+ */
+sealed trait TransactionFilter {
+  def filter(txn: Transaction): Boolean
+  def description(): String
+  def xmlDescription(): xml.Node
+}
+
+/*
+ * Txn:Time filters
+ * To create filter for time span, stack "Begin" and "End" filters.
+ */
+case class EndDateTxnFilter(end: Date) extends TransactionFilter {
+  override def filter(txn: Transaction) = { txn.date.toInt <= end.toInt }
+  override def description() = { "end: Transaction date is on or before: " + end.formatISO8601Ext }
+  override def xmlDescription() = { <filter type="end" date={ end.formatISO8601Ext } />}
```

Comment:
The `description` thing is nice; and doing this with pure functions would have been difficult. So it's good that you stuck to your ground and kept the filter class hierarchy :smile: